src/elf/parser.rs

```diff
@@ -97,24 +90,74 @@ struct Cache {
     sect_cache: Vec<Option<Vec<u8>>>,
 }
 
+impl<'mmap> Cache<'mmap> {
+    /// Create a new `Cache` using the provided raw ELF object data.
+    fn new(elf_data: &'mmap [u8]) -> Self {
+        Self {
+            elf_data,
+            ehdr: None,
+            shdrs: None,
+            shstrtab: None,
+            phdrs: None,
+            symtab: None,
+            symtab_origin: None,
+            strtab: None,
+            str2symtab: None,
+            sect_cache: Vec::new(),
+        }
+    }
+
+    fn ensure_ehdr(&mut self) -> Result<&'mmap Elf64_Ehdr, Error> {
+        match self.ehdr {
```

Comment:
We could parse them eagerly, agreed. But I wouldn't say it buys us a lot (even in terms of complexity reduction), to be honest: instead of making a function call to retrieve the respective item, which may end up "parsing" and caching it or just returning it from a member, we now make a very similar function call during object construction and set a member. The main difference would be that down the line instead of making a function call we'd "just" access a member, but that pretty much makes little to no difference from my perspective.
Then I guess the question is: where do we draw the line? What do we parse eagerly and what not?

Reply:
Up to you, it was just a suggestion. For me field access for stuff that's guaranteed to be there vs Result-returning function (that might fail) is a simplification, but no big deal. As for the drawing the line, I think you can't really do much with ELF without ELF header and section table (which also means having strings section for sections). Wouldn't go much beyond that, because it would be paying upfront costs for something that might not be needed (so e.g., parsing symbol table eagerly would be an obvious step too far, in my book).